.gitignore

```diff
@@ -1,188 +1,8 @@
-*.hex
-
-## Ignore Visual Studio temporary files, build results, and
-## files generated by popular Visual Studio add-ons.
-# User-specific files
-*.suo
-*.user
-*.userosscache
-*.sln.docstates
-# User-specific folders
-*.sln.ide/
-# Build results
-[Dd]ebug/
-[Dd]ebugPublic/
-[Rr]elease/
-[Rr]eleases/
-x64/
-x86/
-bld/
-[Bb]in/
-[Oo]bj/
-.vs/
-# Roslyn cache directories
-*.ide/
-# MSTest test Results
-[Tt]est[Rr]esult*/
-[Bb]uild[Ll]og.*
-#NUNIT
-*.VisualState.xml
-TestResult.xml
-# Build Results of an ATL Project
-[Dd]ebugPS/
-[Rr]eleasePS/
-dlldata.c
-*_i.c
-*_p.c
-*_i.h
-*.ilk
-*.meta
-*.obj
-*.pch
-*.pdb
-*.pgc
-*.pgd
-*.rsp
-*.sbr
-*.tlb
-*.tli
-*.tlh
-*.tmp
-*.tmp_proj
-*.log
-*.vspscc
-*.vssscc
-.builds
-*.pidb
-*.svclog
-*.scc
-# Chutzpah Test files
-_Chutzpah*
-# Visual C++ cache files
-ipch/
-*.aps
-*.ncb
-*.opensdf
-*.sdf
-*.cachefile
-# Visual Studio profiler
-*.psess
-*.vsp
-*.vspx
-# TFS 2012 Local Workspace
-$tf/
-# Guidance Automation Toolkit
-*.gpState
-# ReSharper is a .NET coding add-in
-_ReSharper*/
-*.[Rr]e[Ss]harper
-*.DotSettings.user
-# JustCode is a .NET coding addin-in
-.JustCode
-# TeamCity is a build add-in
-_TeamCity*
-# DotCover is a Code Coverage Tool
-*.dotCover
-# NCrunch
-_NCrunch_*
-.*crunch*.local.xml
-# MightyMoose
-*.mm.*
-AutoTest.Net/
-# Web workbench (sass)
-.sass-cache/
-# Installshield output folder
-[Ee]xpress/
-# DocProject is a documentation generator add-in
-DocProject/buildhelp/
-DocProject/Help/*.HxT
-DocProject/Help/*.HxC
-DocProject/Help/*.hhc
-DocProject/Help/*.hhk
-DocProject/Help/*.hhp
-DocProject/Help/Html2
-DocProject/Help/html
-# Click-Once directory
-publish/
-# Publish Web Output
-*.[Pp]ublish.xml
-*.azurePubxml
-# TODO: Comment the next line if you want to checkin your web deploy settings
-# but database connection strings (with potential passwords) will be unencrypted
-*.pubxml
-*.publishproj
-# NuGet Packages
-*.nupkg
-# The packages folder can be ignored because of Package Restore
-**/packages/*
-# except build/, which is used as an MSBuild target.
-!**/packages/build/
-# If using the old MSBuild-Integrated Package Restore, uncomment this:
-#!**/packages/repositories.config
-# Windows Azure Build Output
-csx/
-*.build.csdef
-# Windows Store app package directory
-AppPackages/
-# Others
-sql/
-*.Cache
-ClientBin/
-[Ss]tyle[Cc]op.*
-~$*
-*~
-*.dbmdl
-*.dbproj.schemaview
-*.pfx
-*.publishsettings
-node_modules/
-bower_components/
-# RIA/Silverlight projects
-Generated_Code/
-# Backup & report files from converting an old project file
-# to a newer Visual Studio version. Backup files are not needed,
-# because we have git ;-)
-_UpgradeReport_Files/
-Backup*/
-UpgradeLog*.XML
-UpgradeLog*.htm
-# SQL Server files
-*.mdf
-*.ldf
-# Business Intelligence projects
-*.rdl.data
-*.bim.layout
-*.bim_*.settings
-# Microsoft Fakes
-FakesAssemblies/
-
-# Xcode project user data
-xcuserdata/
+# VS CMake default output
+/.vs/ProjectSettings.json
+/.vs/*.sqlite
+/.vs/nativefiledialog-extended/
```

Comment:
I'm relatively neutral on this one. I general commit run configurations if that is practical, since that lowers the bar for new people to try the software. Too many projects just don't boot up if you e.g. do not set the correct working directory. This is not the case here, as far as I can tell, so either way is fine with me.

Reply:
On my Windows machine, VS2019 generates 3 non-gitignored files in the `.vs` folder: `\.vs\nativefiledialog\v16\.suo`, `\.vs\nativefiledialog\v16\Browse.VC.db`, `\.vs\VSWorkspaceState.json`.  AFAIK, `Browse.VC.db` is the Intellisense cache and will be automatically generated by VS; and Microsoft says the `.suo` file should not be checked in to source control.

For run configurations, do you mean the things in `CMakeSettings.json` (which is not in the .vs folder)?  That seems to be the proper way when compiling a CMake project in VS2017/2019.

Reply:
Interesting that it generates those files in `\.vs\nativefiledialog` for you - as you might have guessed, those are in `\.vs\nativefiledialog-extended` for me. Did you call your top level project folder `nativefiledialog`? (i.e. the one that you `git clone`d into)

The "run configuration" or "launch configuration" in `launch.vs.json` controls how the resulting binaries are executed when you press the "play" button in VS, i.e. what the current working directory is or what command-line arguments are passed to it. Unfortunately, this rather useful file is in the `/.vs/` folder - I wish it wasn't.

CMakeSettings.json controls which "build types" are available, e.g. Release, Debug, for x86 or X64 etc. It can be useful to share this, but it is usually only really useful if you have a multitude of CMake settings to pass and the defaults don't cut it. 

For simplicities sake, how about I cut the whole `/.vs/` folder and ignore it, as well as the `CMakeSettings.json`?

Reply:
> Interesting that it generates those files in `\.vs\nativefiledialog` for you - as you might have guessed, those are in `\.vs\nativefiledialog-extended` for me. Did you call your top level project folder `nativefiledialog`? (i.e. the one that you `git clone`d into)

Oh yeah I forgot, my top-level directory is `nativefiledialog` because I originally cloned from mlabbe's repo.  Arguably this is easily fixable locally by renaming the top-level directory, but I think it would be rather surprising to others if the name of the local top-level directory influences the way the gitignore works.

> For simplicities sake, how about I cut the whole /.vs/ folder and ignore it

Agreed, I think this is the best option.

> as well as the `CMakeSettings.json`

I don't have a preference for this.  VS2019 will default to a single "x64-Debug" configuration, and developers would have to add any other configurations they need.  This might be nice if they want to add custom compilation options or toolsets.

Reply:
If you think the usefulness of the `launch.vs.json` file justifies an exception to be made for it, I wouldn't mind something like:
```
/.vs/*
!/.vs/launch.vs.json
```

Reply:
Oh nice. I was wondering whether that was possible. Thanks for the info, but let's leave it out for now.